address/address.go

```diff
@@ -47,327 +77,236 @@ const (
 	Testnet
 )
 
-var (
-	// ErrUnknownNetwork is returned when encountering an unknown network in an address.
-	ErrUnknownNetwork = errors.New("unknown network")
-	// ErrUnknownVersion is returned when encountering an unknown address version.
-	ErrUnknownVersion = errors.New("unknown version")
-	// ErrInvalidBytes is returned when encountering an invalid byte format.
-	ErrInvalidBytes = errors.New("invalid bytes")
-)
+// MainnetPrefix is the main network prefix.
+var MainnetPrefix = "f"
 
-// NetworkFromString tries to convert the string representation of a network to
-// its binary representation.
-func NetworkFromString(input string) (Network, error) {
-	switch input {
-	case "fc":
-		return Mainnet, nil
-	case "tf":
-		return Testnet, nil
-	default:
-		return 0, ErrUnknownNetwork
-	}
-}
+// TestnetPrefix is the main network prefix.
+var TestnetPrefix = "t"
 
-// NetworkToString creates a human readable representation of the network.
-func NetworkToString(n Network) string {
-	switch n {
-	case Mainnet:
-		return "fc"
-	case Testnet:
-		return "tf"
-	default:
-		panic("invalid network")
-	}
-}
+// Protocol represents which protocol an address uses.
+type Protocol = byte
 
-// Address is the go type that represents an address in the filecoin network.
-type Address [Length]byte
+const (
+	// ID represents the address ID protocol.
+	ID Protocol = iota
+	// SECP256K1 represents the address SECP256K1 protocol.
+	SECP256K1
+	// Actor represents the address Actor protocol.
+	Actor
+	// BLS represents the address BLS protocol.
+	BLS
+)
 
-// NewMainnet constructs a new mainnet address.
-func NewMainnet(hash []byte) Address {
-	return New(Mainnet, hash)
+// Protocol returns the protocol used by the address.
+func (a Address) Protocol() Protocol {
+	return a.str[0]
 }
 
-// NewTestnet constructs a new testnet address.
-func NewTestnet(hash []byte) Address {
-	return New(Testnet, hash)
+// Payload returns the payload of the address.
+func (a Address) Payload() []byte {
+	return []byte(a.str[1:])
 }
 
-// New constructs a new address for the given nework.
-func New(network Network, hash []byte) Address {
-	var addr [Length]byte
-	addr[0] = network
-	addr[1] = Version
-	copy(addr[2:], hash)
-	return addr
+// Bytes returns the address as bytes.
+func (a Address) Bytes() []byte {
+	return []byte(a.str)
 }
 
-// NewFromString tries to parse a given string into a filecoin address.
-func NewFromString(s string) (Address, error) {
-	networkString, version, hash, err := decode(s)
-	if err != nil {
-		return Address{}, err
-	}
-
-	network, err := NetworkFromString(networkString)
+// String returns an address encoded as a string.
+func (a Address) String() string {
+	str, err := encode(Mainnet, a)
 	if err != nil {
-		return Address{}, err
-	}
-
-	if version != Version {
-		return Address{}, ErrUnknownVersion
+		panic(err)
```

Comment:
This panic is acceptable as long as we can be confident that no address we ever create will trigger it. This panic being hit should be a result of programmer error, and never happen because of anything a user typed or was sent over the network

Reply:
I am fairly confident that this panic will only be caused by a programmer error. `encode()` will will return an error in 2 cases:
1. We pass it an unknown network prefix, this would be a programmer error
2. We pass it an address with an unknown protocol, `decode` will and `NewFromBytes` will prevent addresses from being created with invalid protocols.

It would be great to have fuzz tests on addresses to further guarantee this.

Reply:
This makes me sad, I really don't like returning an error from a `String()` method. I wrote a quick fuzz test for this, and it does indeed panic in our leb128 library, this means returning an error here won't help us. This is the first time I have tried fuzz testing, so I may be doing something wrong here -- the commit with the change is: 82820b656e437bd796967699cf167f4623c62434, failure should be in CI shortly.

Reply:
you really really really should not be panicing on a codepath that could possibly be a function of user input. far better to deal with ugly return signatures or to return a known-bad string or something than to panic. seriously, don't do this, it stops the node from being able to do anything useless and causes the node to exit in an unknown state so you may not ever be able to start your node again. https://github.com/filecoin-project/go-filecoin/blob/master/CONTRIBUTING.md#error-handling

Reply:
Alright, I will return an error. It can be removed later in a follow on if we don't like it (I don't like it, but I like it more than rebasing this branch)

Reply:
This is already what we do for Cids. More worrisome than panicking here is allowing invalid address objects to be created and propogated around the codebase, being sent to other nodes over the network, or even making it into the chain somehow. We should have strict confidence that anything parsing untrusted input catches all cases.

Let's merge this (aside from this discussion, i think we all agree it looks good) and open an issue to fuzz our inputs, for addresses, and for any other places we get random data from places.